helion/_compiler/matmul_utils.py

```diff
@@ -0,0 +1,299 @@
+    )
+
+    # Pad accumulator if needed
+    acc_pad: ast.AST | None = acc
+    if acc and (pad_m or pad_n):
+        assert acc_pad is not None  # We just checked acc is not None
+
+        # Apply padding for each dimension that needs it
+        for pad_flag, dim_val, min_val, pad_dim, orig_dim in [
+            (pad_m, m_val, min_m, 0, m),
+            (pad_n, n_val, min_n, 1, n),
+        ]:
+            if not pad_flag:
+                continue
+
+            assert dim_val is not None
+            factor = min_val // dim_val
+
+            # Determine the other dimension value
+            if pad_dim == 0:  # Padding M
+                other_dim = min_n if pad_n else n
+            else:  # Padding N
+                other_dim = min_m if pad_m else m
+
+            # Build initial shape for reshape
+            if pad_dim == 0:
+                if shape_str_fn and orig_dim and other_dim:
+                    initial_shape = [orig_dim, 1, other_dim]
+                else:
+                    initial_shape = f"[{dim_val}, 1, -1]"
+            else:
+                if shape_str_fn and other_dim and orig_dim:
+                    initial_shape = [other_dim, orig_dim, 1]
+                else:
+                    initial_shape = f"[-1, {dim_val}, 1]"
+
+            # Build mask expression
+            mask_dims = "[None, :, None]" if pad_dim == 0 else "[None, None, :]"
+            mask_expr = f"(tl.arange(0, {factor}) == 0){mask_dims}"
+
+            # Build final shape
+            if pad_dim == 0:
+                final_shape = [min_val, other_dim] if other_dim else f"[{min_val}, -1]"
+            else:
+                final_shape = [other_dim, min_val] if other_dim else f"[-1, {min_val}]"
+
+            acc_pad = _apply_reshape_mask_reshape(
+                acc_pad, initial_shape, mask_expr, final_shape
+            )
+
+    # Perform dot operation
+    if acc_pad:
+        kwargs["acc"] = acc_pad
+    result = emit_tl_dot(lhs_pad, rhs_pad, **kwargs)
+
+    # Unpad result by summing padded dimensions
+    if pad_n:
+        assert n_val is not None  # pad_n is only True if n_val is not None
+        factor = min_n // n_val
+        m_dim = min_m if pad_m else m
+        shape = (
+            shape_str_fn([m_dim, n_val, factor])
+            if shape_str_fn and m_dim
+            else f"[-1, {n_val}, {factor}]"
+        )
+        result = expr_from_string(f"tl.reshape({{x}}, {shape})", x=result)
+        result = expr_from_string("tl.sum({x}, -1)", x=result)
+
+    if pad_m:
+        assert m_val is not None  # pad_m is only True if m_val is not None
+        factor = min_m // m_val
+        n_dim = n
+        shape = (
+            shape_str_fn([m_val, factor, n_dim])
+            if shape_str_fn and n_dim
+            else f"[{m_val}, {factor}, -1]"
+        )
+        result = expr_from_string(f"tl.reshape({{x}}, {shape})", x=result)
+        result = expr_from_string("tl.sum({x}, -2)", x=result)
```

Comment:
Could we use `tl.split`/`tl.join` to generate cleaner code that doesn't need `tl.sum`?

To convert size 8 to 16.
```py
x0.shape == [8, K]
x1 = tl.join(x0, tl.zeros_like(x0))  # doubles the size (all blocks must be powers of 2)
x2 = tl.permute(x1, [0, 2, 1])
x3 = tl.reshape(x2, [16, K])
```

Then in reverse:
```py
y0.shape == [16, K]
y1 = tl.reshape(y0, [8, 2, K])
y2 = tl.permute(y1, [0, 2, 1])
y3, _ = tl.split(y2)
```

Since all blocks are powers of 2 double/half sizes is all you need.